Add a new tutorial which shows how to debug a failing pipeline (#16225)
PrefectHQ/prefect: docs/v3/tutorials/debug.mdx

---
title: Debug a data pipeline

> description: Find out when flow runs go wrong and how to troubleshoot them.

Reviewer: ```suggestion
description: Learn how to troubleshoot flow runs that error.
```

Author: Error is a noun and I think a verb would be more grammatically correct here, so perhaps:

```suggestion
description: Learn how to troubleshoot flow runs that fail.
```

wdyt @discdiver?

---

In [Set up a platform for data pipelines](/v3/tutorials/platform), you used Prefect Cloud to set up a platform for data pipelines.
In this tutorial, you'll learn what to do when those data pipelines fail.

<Info>

> This tutorial starts where the [previous tutorial](/v3/tutorials/platform) leaves off, so complete that one first.

Author: I would love to have a Terraform script they can run to provision Prefect Cloud automatically. I might need some help with this though.

> This tutorial starts where the [previous tutorial](/v3/tutorials/platform) leaves off, so complete that one first.

Reviewer: ooo. That would be cool. Could add as a linear issue.

> This tutorial starts where the [previous tutorial](/v3/tutorials/platform) leaves off, so complete that one first.

Author: https://github.com/PrefectHQ/demos/pull/2

## Deploy a fix

Open the `simulate_failures.py` file and look at line 12.

```python simulate_failures.py {12}
from prefect import flow, task
import argparse
import asyncio
from prefect.client.orchestration import get_client

@task
def process_data(run: int, fail_at_run: int | None = None) -> bool:
    """Simulate data processing with failures"""
    
    # Simulate persistent failures
    if fail_at_run and run > fail_at_run:
        raise Exception(f"Run failed")
    
    return True

# ...
```

The `if` statement is the problem.
If you specify the `--fail_at_run` flag, once the flow runs more than `fail_at_run` times, the flow fails with an exception.
Remove the `if` statement to fix this failure.

```python simulate_failures.py
from prefect import flow, task
import argparse
import asyncio
from prefect.client.orchestration import get_client

@task
def process_data(run: int, fail_at_run: int | None = None) -> bool:
    """Simulate data processing with failures"""
    
    return True

# ...
```

Now, all flow runs succeed in spite of the `--fail-at-run` flag.
Deploy the fix to the staging workspace to confirm this new behavior.

> ```

Reviewer: ```suggestion
```bash
```

prefect cloud workspace --set "<account>/staging"
python simulate_failures.py --fail-at-run 3
```

After the script finishes, open the **Home** page in Prefect Cloud to verify that the flow run is no longer failing.

> You can now deploy the fix to the production workspace as well.

Author: ```suggestion
You can now switch workspaces to update the code used in the production workspace as well.
```
